jax/_src/scipy/sparse/linalg.py

```diff
@@ -123,10 +129,107 @@ def body_fun(value):
   return x_final
 
 
+# aliases for working with pytrees
+
+def _bicgstab_solve(A, b, x0=None, *, maxiter, tol=1e-5, atol=0.0, M=_identity):
+
+  # tolerance handling uses the "non-legacy" behavior of scipy.sparse.linalg.bicgstab
+  bs = _vdot_real_tree(b, b)
+  atol2 = jnp.maximum(jnp.square(tol) * bs, jnp.square(atol))
+
+  # https://en.wikipedia.org/wiki/Biconjugate_gradient_stabilized_method#Preconditioned_BiCGSTAB
+
+  def cond_fun(value):
+    x, r, *_, k = value
+    rs = _vdot_real_tree(r, r)
+    # the last condition checks breakdown
+    return (rs > atol2) & (k < maxiter) & (k >= 0)
+
+  def body_fun(value):
+    x, r, rhat, alpha, omega, rho, p, q, k = value
+    rho_ = _vdot_tree(rhat, r)
+    beta = rho_ / rho * alpha / omega
+    p_ = _add(r, _mul(beta, _sub(p, _mul(omega, q))))
+    phat = M(p_)
+    q_ = A(phat)
+    alpha_ = rho_ / _vdot_tree(rhat, q_)
+    s = _sub(r, _mul(alpha_, q_))
+    exit_early = _vdot_real_tree(s, s) < atol2
+    shat = M(s)
+    t = A(shat)
+    omega_ = _vdot_tree(t, s) / _vdot_tree(t, t)  # make cases?
+    x_ = lax.cond(
+      exit_early,
+      lambda _: _add(x, _mul(alpha_, phat)),
+      lambda _: _add(x, _add(_mul(alpha_, phat), _mul(omega_, shat))),
+      None
+    )
+    r_ = lax.cond(
+      exit_early,
+      lambda _: s,
+      lambda _: _sub(s, _mul(omega_, t)),
+      None
+    )
+    k_ = lax.cond((omega_ == 0) | (alpha_ == 0),
+                  lambda _: -11, lambda _: k + 1, None)
+    k_ = lax.cond((rho_ == 0), lambda _: -10, lambda _: k_, None)
+    return x_, r_, rhat, alpha_, omega_, rho_, p_, q_, k_
+
+  r0 = _sub(b, A(x0))
+  # need to use _vdot_tree to match dtype (hacky...)
```

Comment:
You could tree map `ones_like` I think?

Reply:
Here I am just trying to get the number 1 to match the dtype (complex or float, 32 or 64 bit). How would that work?

Reply:
Sorry... thought I was being helpful with my drive by comment... so you actually want a scalar `1` . Perhaps https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.result_type.html#jax.numpy.result_type would be helpful, but I'll defer to someone else who knows for sure!